### ISSUE TITLE
TTML captions with overlapping times appear weirdly.

### ISSUE THREAD

**Issue by FireMasterK** [google/shaka-player#3186]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
3.0.8

**Can you reproduce the issue with our latest release version?**
Yes

**Can you reproduce the issue with the latest code from `master`?**
Not tried

**Are you using the demo app or your own custom app?**
Custom

**If custom app, can you reproduce the issue using our demo app?**
I have no idea how to load remote captions, but I assume it is

**What browser and OS are you using?**
Windows 10, Brave

**What are the manifest and license server URIs?**

I can send on request.
TTML FIle: https://bin.snopyta.org/?2562701c617b9d7e#3fVSpTeKHn8maiNLGkoQgF4Bty3UjxMEsavUzY8EVfK9

**What did you do?**
Load the specific TTML caption, given above.

**What did you expect to happen?**
Display the caption properly, like how it does on YouTube.

**What actually happened?**
You can see, that it is displaying two different timelines incorrectly.
You can see what happens here: https://piped.kavin.rocks/watch?v=i54eOei2ESk
![image](https://user-images.githubusercontent.com/20838718/109298027-f6eb4100-7858-11eb-9b52-5b57be4b106d.png)
It should display like how it does on YouTube: https://www.youtube.com/watch?v=i54eOei2ESk

**Comment by joeyparrish** [comment]
> You can see what happens here: https://piped.kavin.rocks/watch?v=i54eOei2ESk

I see the problem on that page, but not in our demo app running the same version of Shaka Player.

Here's how I loaded your text content in the v3.0.8 demo:
1. Open the demo: https://shaka-player-demo.appspot.com/demo/#build=compiled
2. Play something
3. Pause it
4. Open the JS debugger
5. Run `player = video.ui.getControls().getLocalPlayer();`
6. Run `player.addTextTrack('https://storage.googleapis.com/shaka-demo-assets/_bugs/3186/subs.ttml', 'xx', 'caption', 'application/ttml+xml');`
7. Select language "xx" from the subtitles UI
8. Unpause the video

Here's what that looks like in action:

https://user-images.githubusercontent.com/6843156/109562245-fdfd9380-7a92-11eb-986b-5d5edb5f5223.mp4

-----

I tried in v3.0.8 in compiled & uncompiled modes, and in the latest master branch version and could not reproduce your issue.  I found that your version is also v3.0.8 by running `document.querySelector('video').ui.getControls().getLocalPlayer().constructor.version` in the debugger on your site.

I first thought you might have some customization or text display plugin or something to get this different behavior, but while investigating that possibility, I spotted something really strange in the DOM on your site:

![Screenshot 2021-03-01 13 36 59 - Display 2](https://user-images.githubusercontent.com/6843156/109562900-dce97280-7a93-11eb-9a5b-fa420a3da217.png)

There are two very strange things going on here.  First, there are two sets of UI divs for most things: `shaka-controls-container`, `shaka-client-side-ad-container`, `shaka-spinner-container`.  Second, although there is a `shaka-text-container` div (and only one), it's empty, even when text is being displayed.

So I suspect that:
1. You may be either mixing DOM-based, declarative setup of the UI with programmatic setup of the UI, or you are double-initializing your UI somehow
2. You are probably seeing the browser's native text display, rather than ours, so the placement issue you see is not in our control

Can you show us how you initialize the Shaka Player UI?

**Comment by joeyparrish** [comment]
Indeed, you can see the browser's native display is being used by opening the video element's shadow root:

![Screenshot 2021-03-01 13 44 38 - Display 2](https://user-images.githubusercontent.com/6843156/109563493-b2e48000-7a94-11eb-9536-92f81212fafc.png)

This all works correctly in our UI, but something about your setup is broken, so our UI is not being used for text.  If you can show us how you're setting it up, we can help you fix it.  Thanks!

**Comment by FireMasterK** [comment]
Hi, thanks for the detailed answer!

Here is how I load the player:
```
// import statements
import "shaka-player/dist/shaka-player.ui.js";
import("shaka-player/dist/controls.css");
const shaka = import("shaka-player/dist/shaka-player.ui.js");
```

```
// relavent code
              player
                .load(
                    "data:application/dash+xml;charset=utf-8;base64," +
                        btoa(dash)
                )
                .then(() => {
                    player.getVariantTracks().forEach(track => {
                        console.log(track.height);
                    });
                    this.video.subtitles.map(subtitle => {
                        player.addTextTrack(
                            subtitle.url,
                            "eng",
                            "SUBTITLE",
                            subtitle.mimeType,
                            null,
                            "English"
                        );
                        player.setTextTrackVisibility(true);
                    });
                    if (localStorage)
                        videoEl.volume = localStorage.getItem("volume") || 1;

                    const ui = // done so that I don't create 2 UIs
                        window.ui ||
                        (window.ui = new shaka.ui.Overlay(
                            player,
                            document.querySelector(
                                "div[data-shaka-player-container]"
                            ),
                            videoEl
                        ));
                    const config = {
                        overflowMenuButtons: [
                            "quality",
                            "captions",
                            "playback_rate"
                        ]
                    };
                    ui.configure(config);
                });
```

I am aware of `video["ui"]`, but I always have a race condition when using it :/

**Comment by joeyparrish** [comment]
There's a lot of strange things going on here.  First, I don't know how your import statements work, but my instinct says you shouldn't have two imports of the same library.  This may create a setup race.

Second, you create a UI Overlay explicitly using `document.querySelector('div[data-shaka-player-container]')`, but that doesn't make sense.  `div[data-shaka-player-container]` is used by the library to create the UI for you.  If you have that in your HTML, you would never create the Overlay explicitly.

If you use DOM-based setup, use that **only**.  If you use the API for setup, use that **only**.

I recommend you stick to DOM-based setup.  What you're probably missing that is causing your race on video.ui is that you need to listen for an event we fire to tell you when the UI setup is complete.  For example, see our [UI tutorial](https://shaka-player-demo.appspot.com/docs/api/tutorial-ui.html), which says:

```js
// Listen to the custom shaka-ui-loaded event, to wait until the UI is loaded.
document.addEventListener('shaka-ui-loaded', init);
// Listen to the custom shaka-ui-load-failed event, in case Shaka Player fails
// to load (e.g. due to lack of browser support).
document.addEventListener('shaka-ui-load-failed', initFailed);
```

So, to summarize, I suggest you:
1. Only import the library once
2. Wait for the `shaka-ui-loaded` event
3. Get UI from `video.ui` and the player from `ui.getControls().getPlayer()`

Does this help?

**Comment by FireMasterK** [comment]
The problem I have with 
```
document​.​addEventListener​(​'shaka-ui-loaded'​,​ ​init​)​;
```

Is that sometimes I register this listener a lot later, thus the event not firing, is there any other way to solve this?

**Comment by FireMasterK** [comment]
Actually, the approach which I've seems to be fine according to the docs since I want a `Programmatic UI setup` as I want to set up a player after the page loads.

Another thing: This issue does not occur on Firefox, which is even more strange

**Comment by joeyparrish** [comment]
> The problem I have with
> 
> ```
> document​.​addEventListener​(​'shaka-ui-loaded'​,​ ​init​)​;
> ```
> 
> Is that sometimes I register this listener a lot later, thus the event not firing, is there any other way to solve this?

I believe you can check if `video.ui` is present, and if so, the event has already fired.  We should probably consider something more formal, though.

> Actually, the approach which I've seems to be fine according to the docs since I want a Programmatic UI setup as I want to set up a player after the page loads.

Oh, good.  Yes, that's preferable when using dynamic UI frameworks.

Does this answer all your questions?

**Comment by FireMasterK** [comment]
Closing as I managed to fix this, when I moved my player to a new component, everything fixed itself.
